changelog/ob.md

```diff
@@ -0,0 +1,296 @@
+# Prototype Ownership/Borrowing System for Pointers
+
+An Ownership/Borrowing (aka OB) system for pointers can guarantee that
+dereferenced pointers are pointing to a valid memory object.
+
+
+## Scope of Prototype OB System
+
+This is a prototype OB system adapted to D. It is initially for pointers
+only, not dynamic arrays, class references, refs, or pointer fields
+of aggregates. Adding support for such adds complexity,
+but does not change the nature of it, hence it is deferred to later.
+RAII objects can safely manage their own memory, so are not covered
+by OB. Whether a pointer is allocates memory using the GC or some other
+storage allocator is immaterial to OB, they are not distinguished and are
+handled identically.
+
+The system is only active in functions annotated with the `@live` attribute.
+It is applied after semantic processing is done as purely a check
+for violations of the OB rules. No new syntax is added. No change is made
+to the code generated.
+If `@live` functions
+call non-`@live` functions, those called functions are expected to present
+an `@live` compatible interface, although it is not checked.
```

Comment:
Given:

```d
void nonlive(int *p);
void fun() @live {
    auto p = cast(int*) malloc(int.sizeof);
    nonlive(p);
}
```

Would it be the case then under this proposal that `nonlive(p)` is considered to "consume" p?

Reply:
That was my understanding. Which makes it work with common approaches like "Library_init" and "Library_free". On the other hand, it does require a lot of annotations to work with *anything else*. And some use cases (like `getopt`) are probably not going to work (anything that requires multiple mutable pointers).

Reply:
Yes, exactly.

> it does require a lot of annotations to work with anything else. 

Don't see that.

> some use cases (like getopt) are probably not going to work (anything that requires multiple mutable pointers).

That's well known with Rust. They struggle with it a lot, which is why D allows you to dip in and out of @live as required.